akka-http-core/src/main/scala/akka/http/scaladsl/model/ServerSentEvent.scala

```diff
@@ -0,0 +1,143 @@
+import scala.compat.java8.OptionConverters.RichOptionForJava8
+
+object ServerSentEvent {
+
+  /**
+   * A [[ServerSentEvent]] with empty data which can be used as a heartbeat.
+   */
+  val heartbeat: ServerSentEvent =
+    ServerSentEvent("")
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param type type, must not contain \n or \r
+   */
+  def apply(data: String, `type`: String): ServerSentEvent =
+    new ServerSentEvent(data, Some(`type`))
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param type type, must not contain \n or \r
+   * @param id id, must not contain \n or \r
+   */
+  def apply(data: String, `type`: String, id: String): ServerSentEvent =
+    new ServerSentEvent(data, Some(`type`), Some(id))
+
+  /**
+   * Creates a [[ServerSentEvent]].
+   *
+   * @param data data, may span multiple lines
+   * @param retry reconnection delay in milliseconds
+   */
+  def apply(data: String, retry: Int): ServerSentEvent =
+    new ServerSentEvent(data, retry = Some(retry))
+
+  private def noNewLine(s: String) = s.forall(c ⇒ c != '\n' && c != '\r')
+
+  // Public domain algorithm: http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2.
+  // We want powers of two both because they typically work better with the allocator, and because we want to minimize
+  // reallocations/buffer growth.
+  private def nextPowerOfTwoBiggerThan(n: Int) = {
+    var m = n - 1
+    m |= m >> 1
+    m |= m >> 2
+    m |= m >> 4
+    m |= m >> 8
+    m |= m >> 16
+    m + 1
+  }
+}
+
+/**
+ * Representation of a server-sent event. According to the specification, an empty data field designates an event
+ * which is to be ignored which is useful for heartbeats.
+ *
+ * @param data data, may span multiple lines
+ * @param type optional type, must not contain \n or \r
+ * @param id optional id, must not contain \n or \r
+ * @param retry optional reconnection delay in milliseconds
+ */
+final case class ServerSentEvent(
+  data:   String,
+  `type`: Option[String] = None,
+  id:     Option[String] = None,
+  retry:  Option[Int]    = None)
+  extends javadsl.model.ServerSentEvent {
+  import ServerSentEvent._
+
+  require(`type`.forall(noNewLine), "type must not contain \\n or \\r!")
+  require(id.forall(noNewLine), "id must not contain \\n or \\r!")
+  require(retry.forall(_ > 0), "retry must be a positive number!")
+
+  private[scaladsl] def encode = {
+    def s = { // Performance fun fact: change this to val and get an ~30% performance penalty!!!
+      // Why 8? "data:" == 5 + \n\n (1 data (at least) and 1 ending) == 2 and then we add 1 extra to allocate
+      //        a bigger memory slab than data.length since we're going to add data ("data:" + "\n") per line
+      // Why 7? "event:" + \n == 7 chars
+      // Why 4? "id:" + \n == 4 chars
+      // Why 17? "retry:" + \n + Integer.Max decimal places
```

Comment:
Could be assigned into local vals and commented like this; good comments though, thanks